yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+        logDebug(s"${eventsProcessed} events are processed")
+      }
+      val sparkEvent = event.sparkEvent
+      sparkEvent match {
+        case start: SparkListenerApplicationStart =>
+          // we already have all information,
+          // flush it for old one to switch to new one
+          logDebug(s"Handling application start event: $event")
+          appName = start.appName
+          userName = start.sparkUser
+          startTime = start.time
+          if (startTime == 0) {
+            startTime = now()
+          }
+          setContextAppAndAttemptInfo(
+            start.appId,
+            start.appAttemptId)
+          // flush old entity
+          resetCurrentEntity
+          appStartEventProcessed.set(true)
+          appEndEventProcessed.set(false)
+          // force create the new one
+          val en = started(getOrCreateCurrentEntity, startTime)
+          logInfo(s"Application started: $event")
+          push = true
+
+
+        case end: SparkListenerApplicationEnd =>
+          if (!appEndEventProcessed.get()) {
+            // we already have all information,
+            // flush it for old one to switch to new one
+            logInfo(s"Application end event: $event")
+            if (!appStartEventProcessed.get()) {
+              logWarning(s"Handling application end event without application start $event")
+            }
+            // flush old entity
+            val endtime = if (end.time > 0) end.time else now()
+            if (startTime == 0) {
+              // special case: an app end is received without any active start
+              // event sent yet
+              logInfo(s"start time is 0; fixing up to match end time")
+              startTime = endtime
+            }
+            // reset the current entity
+            resetCurrentEntity
+            // then trigger its recreation
+            val en = completed(getOrCreateCurrentEntity(),
+              startTime,
+              endtime,
+              sparkApplicationId,
+              sparkApplicationAttemptId)
+            appStartEventProcessed.set(false)
+            appEndEventProcessed.set(true)
+            push = true
+          }
+        case _ =>
+          val currentEntity: TimelineEntity = getOrCreateCurrentEntity()
+      }
+
+      val tlEvent = toTimelineEvent(event)
+      getOrCreateCurrentEntity().addEvent(tlEvent)
+      logDebug(s"Event: $event => ${describeEntity(getOrCreateCurrentEntity())}")
+
+      currentEventCount += 1
+      // set the push flag if the batch limit is reached
+      push |= currentEventCount == batchSize
+      // after the processing, increment the counter
+      // this is done at the tail to count real number of processed events, not count received
+      incEventsProcessed
+    } // end of synchronized clause
+
+    logDebug(s"current event num: $currentEventCount")
+    if (push) {
+      logDebug("Push triggered")
+      flushCurrentEventsToATS
```

Comment:
this is midly confusing because this method calls `flushCurrentEventsToATS` which in turn calls this method again... it's safe because of the batch handling, but it would feel cleaner to call `asyncFlush` instead.